hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java

```diff
@@ -0,0 +1,163 @@
+        maxRateRegistry.start();
+    }
+
+    @After
+    public void cleanup() throws Exception {
+        deleteAllNodes();
+        maxRateRegistry.stop();
+    }
+
+    @Test
+    public void shouldHaveEmptyRateHistoryOnInit() {
+        // given
+        maxRateRegistry.onBeforeMaxRateCalculation(); // read
+
+        // and
+        maxRateRegistry.ensureCorrectAssignments(subscription1, ImmutableSet.of(consumerId));
+
+        // then
+        assertEquals(RateHistory.empty(), maxRateRegistry.getRateHistory(new ConsumerInstance(consumerId, subscription1)));
+    }
+
+    @Test
+    public void shouldWriteAndReadRateHistoryForCurrentConsumer() {
+        // given
+        ConsumerInstance consumer = new ConsumerInstance(consumerId, subscription1);
+
+        RateHistory rateHistory = RateHistory.create(0.5);
+
+        // when
+        maxRateRegistry.writeRateHistory(consumer, rateHistory);
+
+        // and
+        maxRateRegistry.onAfterWriteRateHistories(); // store
+
+        // then
+        wait.untilZookeeperPathIsCreated(paths.currentConsumerRateHistoryPath());
+
+        // when
+        maxRateRegistry.onBeforeMaxRateCalculation(); // read
+
+        // and
+        RateHistory readHistory = maxRateRegistry.getRateHistory(consumer);
+
+        // then
+        assertEquals(rateHistory, readHistory);
+    }
+
+    @Test
+    public void shouldWriteAndReadMaxRateForCurrentConsumer() {
+        // when
+        maxRateRegistry.update(subscription1, ImmutableMap.of(consumerId, new MaxRate(0.5)));
+        maxRateRegistry.update(subscription2, ImmutableMap.of(consumerId, new MaxRate(350.0)));
+
+        // and
+        maxRateRegistry.onAfterMaxRateCalculation(); // store
+
+        // then
+        wait.untilZookeeperPathIsCreated(paths.consumerMaxRatePath(consumerId));
+
+        // when
+        maxRateRegistry.onBeforeMaxRateCalculation(); // read
+
+        // then
+        assertEquals(new MaxRate(0.5), maxRateRegistry.getMaxRate(new ConsumerInstance(consumerId, subscription1)).get());
+        assertEquals(new MaxRate(350.0), maxRateRegistry.getMaxRate(new ConsumerInstance(consumerId, subscription2)).get());
+    }
+
+    @Test
+    public void shouldCleanRegistryFromInactiveConsumerNodes() {
+        // when
+        maxRateRegistry.update(subscription1, ImmutableMap.of(consumerId, new MaxRate(350.0)));
+
+        // and
+        maxRateRegistry.onAfterMaxRateCalculation(); // store
+
+        // then
+        wait.untilZookeeperPathIsCreated(paths.consumerMaxRatePath(consumerId));
+
+        // when
+        when(assignedConsumersSupplier.getAllAssignedConsumers()).thenReturn(Collections.emptySet());
+        maxRateRegistry.onBeforeMaxRateCalculation(); // cleanup
+
+        // then
+        wait.untilZookeeperPathNotExists(paths.consumerMaxRatePath(consumerId));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void shouldNotAllowForSavingRateHistoryOfOtherNodes() {
```

Comment:
What `other` means in this context? I think it's not clear.

Reply:
Here it is possible to get maximum rate or rate history only for the current node on which the method is called, it is not possible to get data related to "other" nodes in the cluster. So the leader making the calculation won't be able to fetch data using this method. That's why I've added some preconditions in order not to make a mistake in using it. 

I'm aware the interface is not very clear unfortunately, but I wanted to stick as much as possible to the previous max-registry implementation and not refactor everything.

Reply:
Thanks for the explanation.